index.md

```diff
@@ -1,16 +1,128 @@
+
+1. The payload is pip-installable: there's a `requirements.txt` file, the installer points `pip` at it, and the payload is installed.
+2. The payload is conda-installable: there's an environment `YAML` file, the installer points `conda` at it, and the payload is installed.
+3. The payload is supplied as a Dockerfile that constructs a container image.  The appropriate pieces of the Dockerfile can be moved into the nublado-launchable image construction (either as `COPY` and `RUN` statements, or as a shell script which does the same thing), and the payload is installed.
+4. The payload is only available as a Docker image, without build instructions.  This is alarming and certainly suggests the first method would be a better idea, but assuming that the payload is limited somehow to a particular place in the filesystem (for instance, the DM pipelines stack lives almost entirely under `/opt/lsst/software/stack`), a multi-stage Docker build could transplant that directory hierarchy into the target container.
+5. The payload is delivered as an installation script.  The DM pipelines stack can also be installed this way with [lsstinstall][2] and this may be the most common way for thorny environments to be delivered.  For anything of sufficent size, that has made the (bad) choice to not use a standard build/installation system, this will probably already exist.  See, for instance, the huge proliferation in products that expect you to `curl https://some-url | sudo /bin/bash`.  Less grotesquely, maybe it's something you can install with `./configure; make; make install`.
+
+My preference is the second method.  That lets us keep almost all of the image-building machinery the same between payloads, and by encapsulating the payload installation in a script and image layer of its own, replacement of the payload becomes quite straightforward.  If the payload is in one of the easy formats, this script might literally be nothing more than `pip install -r requirements.txt`.  However, specific image-building considerations mean that it may not be a single script: system-level packages may need to be installed as root, while we would prefer that both the installation and runtime operation of the payload happen as a less-privileged user.
+
+That sets up the third point: what will be necessary in order to turn the installed payload into a Jupyterlab-runnable kernel and a terminal environment with the payload contents configured and available?
+
+## Payload configuration
+
+Configuration of the payload environment has at least three components.
+
+Setting up the Jupyter kernel is the most straightforward.  If the payload can be represented in a Python virtual environment, `python -m ipykernel install` is all that needs to be done.  If the payload is more complicated (as the DM pipelines stack is), then it turns out that the kernel can be a shell script that does necessary setup and execs a python process as its last action.
```

Comment:
I would be explicit that the provider of the payload needs to provide this script and the bash shell machinery in the next item. (Exactly how is probably a matter for the Nublado documentation.)